Imgix Integration

cawfree opened the pull request:
### Expected Behaviour 
  - If `IMGIX_DOMAIN` and `IMGIX_TOKEN` are not specified, the app should continue with a warning that image signing is disabled.
  - Else, all network images should be signed via Imgix. (Included `preload`ed images)
  - All image content should render correctly. Static resources (i.e. local assets) should _not_ be signed even when rendered, or loaded, using `ImgixImage`.

### Changes
  - Migrated all usage of `FastImage` to `ImgixImage`.
    - This exports an identical interface, but just conditionally manages signing.
  - Updated `README.md` and `.env.example` to respect the presence of `IMGIX_DOMAIN` and `IMGIX_TOKEN` variables.
  - Included image signing to all components which attempt to render `FastImage`, `RN.Image`, `RN.Animated.Image` and `react-primitives.Image`.
  - All image urls that are passed into native components, i.e. `NativeModules.TransactionList` will also be signed.

### Further Enhancements
  - Currently we are storing all image data in an unbounded in-memory cache to reduce the number of attempts made to sign a network URL. We should consider using an [LRU](https://yomguithereal.github.io/mnemonist/lru-cache.html) to place a cap on maximum memory consumption.

osdnk commented on src/components/buttons/rainbow-button/RainbowButton.js:
```diff
@@ -1,8 +1,8 @@
-import FastImage from 'react-native-fast-image';
 import styled from 'styled-components/primitives';
 import AddCashIconSource from '../../../assets/addCashIcon.png';
+import ImgixImage from '../../ImgixImage';
```
Can you use some alias here like `@images/ImgixImage`

cawfree replied:
Thanks for checking it out 🙏 That sounds great!

Do you think it would make sense to have a dedicated directory for image-style components like:

```diff
"@rainbow-me/apollo": ["./src/apollo"],
"@rainbow-me/assets": ["./src/assets"],
"@rainbow-me/config": ["./src/config"],
+ "@rainbow-me/images": ["./src/components/images"],
```

This way we could probably pull over `ImgixImage`, `ImageWithCachedMetadata` `Avatar` and `ImageAvatar`, as these seem the most used by the wider project. Then we could use:

```typescript
import { ImgixImage, ImageAvatar, ImageWithCachedMetadata, Avatar } from "@rainbow-me/images";
```

Or would it be preferable to use a direct path?

brunobar79 replied:
Sounds good to me

brunobar79 commented on src/handlers/imgix.ts:
```diff
@@ -0,0 +1,107 @@
+    // will not exist if the .env hasn't been configured correctly.
+    if (staticImgixClient) {
+      // eslint-disable-next-line no-console
+      __DEV__ && console.log(`[Imgix]: Signed "${externalImageUri}".`);
```
We don't need this log.

brunobar79 commented on src/handlers/imgix.ts:
```diff
@@ -0,0 +1,107 @@
+    }
+  } catch (e) {
+    // eslint-disable-next-line no-console
+    console.log(`[Imgix]: Failed to sign "${externalImageUri}"!`, e);
```
use logger.log

brunobar79 commented on src/handlers/imgix.ts:
```diff
@@ -0,0 +1,107 @@
+    });
+  }
+  // eslint-disable-next-line no-console
+  console.log(
```
logger.log

brunobar79 commented on src/handlers/imgix.ts:
```diff
@@ -0,0 +1,107 @@
+      // eslint-disable-next-line no-console
+      __DEV__ && console.log(`[Imgix]: Signed "${externalImageUri}".`);
+      return staticImgixClient.buildURL(externalImageUri, {});
+    }
```
if there's no static client we should return the original url, otherwise images won't work for open source contributors

cawfree replied:
That _should_ be okay! If the client is available, it should fall back to the default:

```typescript
const shouldSignUri = (externalImageUri: string): string | undefined => {
  try {
    // We'll only attempt to sign if there's an available client. A client
    // will not exist if the .env hasn't been configured correctly.
    if (staticImgixClient) {
      // eslint-disable-next-line no-console
      __DEV__ && console.log(`[Imgix]: Signed "${externalImageUri}".`);
      return staticImgixClient.buildURL(externalImageUri, {});
    }
  } catch (e) {
    // eslint-disable-next-line no-console
    console.log(`[Imgix]: Failed to sign "${externalImageUri}"!`, e);
    // If something goes wrong, it is not safe to assume the image is valid.
    return undefined;
  }
  return externalImageUri;
};
```

brunobar79 commented on src/handlers/imgix.ts:
```diff
@@ -0,0 +1,107 @@
+      return typeof host === 'string' && !!host.length;
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.log(`[Imgix]: Failed to parse "${externalImageUri}"!`, e);
```
logger.log

cawfree replied:
Kicking myself for not thinking to do this. Thanks dude.

brunobar79 reviewed (changes requested):
Looking good!  Added some minor comments

jinchung commented:
Nice! LGTM. How soon after can we get the LRU cache in?

jinchung commented on src/handlers/imgix.ts:
```diff
@@ -0,0 +1,120 @@
+// It should be a non-empty string which points to a remote address.
+// Other strings (such as those which point to local assets) should
+// not be signed.
+export const isPossibleToSignUri = (
```
Doesn't look like this needs to be exported

cawfree replied:
I thought it might be useful as a helper method, but it didn't turn out this way. I'll happily make this private. 🕵️‍♂️

jinchung commented on src/hooks/useImgix.ts:
```diff
@@ -0,0 +1,21 @@
+  }, [maybeUnsafeUri]);
+}
+
+export default function useImgix(source: Source): Source {
```
Doesn't look like this particular default function is used currently, but we're keeping it for potential future use?

cawfree replied:
Hey Jin! Amazing catch 💪, this should 100% be used by `ImgixImage.ts`. D'oh!

```typescript
class ImgixImage extends React.PureComponent<ImgixImageProps> {
  render() {
    // Why am I not doing anything with the source prop?! 🙈
    const { Component: maybeComponent, source, ...props } = this.props;
    const Component = maybeComponent || FastImage;
    return <Component {...props} source={source} />;
  }
}
```

I think I had to remove the hook when I migrated over to using a class component, somewhere during testing it never made it back in. I am a fool.

Even after the update, you're right, this function is unused... I see no harm in removing it since it doesn't look like there's any existing use case for it. I'll fix this ASAP.

jinchung reviewed (approved):
🥑

cawfree commented:
> Nice! LGTM. How soon after can we get the LRU cache in? @cawfree

Thanks so much for the review, @jinchung! I can add this in straight away, I was just unsure about using `mnemonist/lru` in case that there's an existing dependency of Rainbow which might be more suitable to use.

Do you have an idea of what a suitable size limit would be?

brunobar79 commented:
@jinchung @cawfree I suggest we monitor the cache size first along a session before doing any work.

There are simpler things we can do like have a fixed size or wiping / trimming cache when resuming from background (but we need to know how big it is first)

cawfree commented:
> @jinchung @cawfree I suggest we monitor the cache size first along a session before doing any work.
> 
> There are simpler things we can do like have a fixed size or wiping / trimming cache when resuming from background (but we need to know how big it is first)

Great idea 💡, would you like me to occasionally log the cache size? (Maybe when backgrounded?)

brunobar79 commented:
That would be cool. We can remove the log once we take action

cawfree commented:
@brunobar79 @jinchung 
  - Corrected `ImgixImage.ts` to sign the `source` prop in both the `constructor` and `componentDidUpdate` (if the `source` prop has changed and it looks signable).
  - Added a dependency on [`mnemonist/lru`](https://yomguithereal.github.io/mnemonist/lru-cache.html#size) and asserted a default fixed capacity of `256`.
    - This will allocate 256 empty spaces on allocation.
  - Added logging of the`staticSignatureLru` metrics in `App.tsx` when `AppState` moves to `"background"`.

skylarbarrera commented:
iOS looks good 2 go, all images and everything seems to be working as expected.

Android has the following regression, all image assets seem to be working correctly otherwise: http://cloud.skylarbarrera.com/Screen-Shot-2021-01-07-13-30-24.98.png

@mikedemarais or @osdnk may be able to help you with this one

brunobar79 commented:
@cawfree The bug reported earlier by  was not related to this PR so you can ignore.  I'll review one more time tomorrow and it should be good to go

brunobar79 reviewed (approved):
This is good to go  🚀  

 I'll take care of merging this later today because there's another release that I want to make first.

cawfree commented:
@brunobar79 Many thanks 🙏
